File: Tests/CodeRimTests/ParserTests/UsageNormalizerTests.swift
import Foundation
import XCTest
@testable import CodeRim

final class UsageNormalizerTests: XCTestCase {
    private let normalizer = UsageNormalizer()
    private let timestamp = Date(timeIntervalSince1970: 1_800_000_000)

    func testUsesCumulativeIncreaseAndIgnoresRepeatedSnapshot() {
        let first = normalizer.normalize(
            observation(cumulative: usage(100, cached: 60, output: 20)),
            metadata: rootMetadata,
            state: .empty
        )
        XCTAssertEqual(first.delta, usage(100, cached: 60, output: 20))

        let repeated = normalizer.normalize(
            observation(cumulative: usage(100, cached: 60, output: 20)),
            metadata: rootMetadata,
            state: first.state
        )
        XCTAssertNil(repeated.delta)

        let increased = normalizer.normalize(
            observation(cumulative: usage(130, cached: 80, output: 25)),
            metadata: rootMetadata,
            state: repeated.state
        )
        XCTAssertEqual(increased.delta, usage(30, cached: 20, output: 5))
    }

    func testForkedSessionCountsFreshCounterWhenLastMatchesCumulative() {
        let metadata = SessionMetadata(
            id: "child",
            model: nil,
            workingDirectory: nil,
            forkedFromID: "parent"
        )
        let result = normalizer.normalize(
            observation(cumulative: usage(5_000, cached: 4_000, output: 500)),
            metadata: metadata,
            state: .empty
        )

        XCTAssertEqual(result.delta, usage(5_000, cached: 4_000, output: 500))
        XCTAssertEqual(result.state.quality, .exact)
        XCTAssertNil(result.diagnostic)
    }

    func testForkedSessionSkipsAmbiguousInheritedBaseline() {
        let metadata = SessionMetadata(
            id: "child",
            model: nil,
            workingDirectory: nil,
            forkedFromID: "parent"
        )
        let result = normalizer.normalize(
            CodexTokenObservation(
                occurredAt: timestamp,
                ordinal: nil,
                lastUsage: usage(100, cached: 50, output: 20),
                cumulativeUsage: usage(5_000, cached: 4_000, output: 500)
            ),
            metadata: metadata,
            state: .empty
        )

        XCTAssertNil(result.delta)
        XCTAssertEqual(result.state.quality, .partial)
        XCTAssertEqual(result.diagnostic, "initial cumulative baseline is unresolved")
    }

    func testDecreaseDoesNotBecomeFreshUsage() {
        let prior = UsageNormalizationState(
            cumulativeHighWaterMark: usage(1_000, cached: 700, output: 100),
            quality: .exact
        )
        let result = normalizer.normalize(
            CodexTokenObservation(
                occurredAt: timestamp,
                ordinal: 2,
                lastUsage: usage(100, cached: 80, output: 10),
                cumulativeUsage: usage(500, cached: 300, output: 50)
            ),
            metadata: rootMetadata,
            state: prior
        )

        XCTAssertNil(result.delta)
        XCTAssertEqual(result.state.cumulativeHighWaterMark, prior.cumulativeHighWaterMark)
        XCTAssertEqual(result.state.quality, .partial)
    }

    func testValidatedCounterRestartCountsFreshSegmentAsPartial() {
        let prior = UsageNormalizationState(
            cumulativeHighWaterMark: usage(1_000, cached: 700, output: 100),
            quality: .exact
        )
        let fresh = usage(50, cached: 20, output: 10)
        let observation = CodexTokenObservation(
            occurredAt: timestamp,
            ordinal: 2,
            lastUsage: fresh,
            cumulativeUsage: fresh
        )
        let result = normalizer.normalize(observation, metadata: rootMetadata, state: prior)

        XCTAssertEqual(result.delta, fresh)
        XCTAssertEqual(result.state.cumulativeHighWaterMark, fresh)
        XCTAssertEqual(result.state.quality, .partial)
        XCTAssertEqual(result.diagnostic, "cumulative counter restarted")
    }

    func testOlderReplayCannotRegressCurrentSegment() {
        let latestDate = timestamp.addingTimeInterval(60)
        let prior = UsageNormalizationState(
            cumulativeHighWaterMark: usage(150, cached: 90, output: 30),
            lastObservedAt: latestDate,
            quality: .exact
        )
        let replay = CodexTokenObservation(
            occurredAt: timestamp,
            ordinal: nil,
            lastUsage: usage(100, cached: 60, output: 20),
            cumulativeUsage: usage(100, cached: 60, output: 20)
        )

        let result = normalizer.normalize(replay, metadata: rootMetadata, state: prior)

        XCTAssertNil(result.delta)
        XCTAssertEqual(result.state, prior)
        XCTAssertEqual(result.diagnostic, "out-of-order token snapshot ignored")
    }

    func testUnresolvedInitialBaselineIsNotCounted() {
        let cumulative = usage(500, cached: 300, output: 50)
        let observation = CodexTokenObservation(
            occurredAt: timestamp,
            ordinal: 1,
            lastUsage: usage(100, cached: 80, output: 10),
            cumulativeUsage: cumulative
        )
        let result = normalizer.normalize(observation, metadata: rootMetadata, state: .empty)
        XCTAssertNil(result.delta)
        XCTAssertEqual(result.state.quality, .partial)
        XCTAssertEqual(result.diagnostic, "initial cumulative baseline is unresolved")
    }

    func testLastUsageAloneIsNotAcceptedAsAnAuthoritativeDelta() {
        let observation = CodexTokenObservation(
            occurredAt: timestamp,
            ordinal: 1,
            lastUsage: usage(100, cached: 80, output: 20),
            cumulativeUsage: nil
        )
        let result = normalizer.normalize(observation, metadata: rootMetadata, state: .empty)
        XCTAssertNil(result.delta)
        XCTAssertEqual(result.state.quality, .partial)
    }

    private var rootMetadata: SessionMetadata {
        SessionMetadata(id: "root", model: nil, workingDirectory: nil)
    }

    private func observation(cumulative: TokenUsage) -> CodexTokenObservation {
        CodexTokenObservation(
            occurredAt: timestamp,
            ordinal: 1,
            lastUsage: cumulative,
            cumulativeUsage: cumulative
        )
    }

    private func usage(_ input: Int64, cached: Int64, output: Int64) -> TokenUsage {
        TokenUsage(inputTokens: input, cachedInputTokens: cached, outputTokens: output)
    }
}

File: Windows/tests/CodeRim.Core.Tests/UsageNormalizerTests.cs
using CodeRim.Core.Domain;
using CodeRim.Core.Services;
using System.Globalization;

namespace CodeRim.Core.Tests;

public sealed class UsageNormalizerTests
{
    private readonly DateTimeOffset timestamp = DateTimeOffset.Parse(
        "2026-08-27T01:02:03Z",
        CultureInfo.InvariantCulture);

    [Fact]
    public void TotalCountsCachedInputOnlyAsPartOfInput()
    {
        var usage = new TokenUsage(1_200, 800, 300);

        Assert.Equal(1_500, usage.TotalTokens);
    }

    [Fact]
    public void UsesCumulativeIncreaseAndIgnoresRepeatedSnapshot()
    {
        var first = UsageNormalizer.Normalize(Observation(new TokenUsage(100, 60, 20)), UsageNormalizationState.Empty);
        var repeated = UsageNormalizer.Normalize(Observation(new TokenUsage(100, 60, 20)), first.State);
        var increased = UsageNormalizer.Normalize(Observation(new TokenUsage(130, 80, 25)), repeated.State);

        Assert.Equal(new TokenUsage(100, 60, 20), first.Delta);
        Assert.Null(repeated.Delta);
        Assert.Equal(new TokenUsage(30, 20, 5), increased.Delta);
    }

    [Fact]
    public void DoesNotGuessAnAmbiguousInitialBaseline()
    {
        var observation = new TokenObservation(
            timestamp,
            1,
            new TokenUsage(100, 80, 10),
            new TokenUsage(500, 300, 50));

        var result = UsageNormalizer.Normalize(observation, UsageNormalizationState.Empty);

        Assert.Null(result.Delta);
        Assert.Equal(DataQuality.Partial, result.State.Quality);
    }

    [Fact]
    public void CountsAValidatedCounterRestartAsPartial()
    {
        var previous = new UsageNormalizationState(
            new TokenUsage(1_000, 700, 100),
            timestamp.AddMinutes(-1),
            DataQuality.Exact);
        var fresh = new TokenUsage(50, 20, 10);

        var result = UsageNormalizer.Normalize(
            new TokenObservation(timestamp, 2, fresh, fresh),
            previous);

        Assert.Equal(fresh, result.Delta);
        Assert.Equal(DataQuality.Partial, result.State.Quality);
    }


    [Fact]
    public void MissingOptionalCacheWriteDoesNotLoseOrReplayTokens()
    {
        var firstUsage = new TokenUsage(100, 0, 20, 40);
        var first = UsageNormalizer.Normalize(Observation(firstUsage), UsageNormalizationState.Empty);
        var repeated = UsageNormalizer.Normalize(Observation(new(100, 0, 20)), first.State);
        Assert.Null(repeated.Delta);
        var second = UsageNormalizer.Normalize(new(timestamp.AddSeconds(1), 2, new(100, 0, 20), new(200, 0, 40)), first.State);
        var third = UsageNormalizer.Normalize(new(timestamp.AddSeconds(2), 3, new(100, 0, 20), new(300, 0, 60)), second.State);
        Assert.Equal(360, first.Delta!.Value.TotalTokens + second.Delta!.Value.TotalTokens + third.Delta!.Value.TotalTokens);
        Assert.Null(second.Delta.Value.CacheWriteInputTokens);
        Assert.Equal(DataQuality.Exact, third.State.Quality);
    }
    private TokenObservation Observation(TokenUsage usage) =>
        new(timestamp, 1, usage, usage);
}
